src/service/deferred-builder.js

```diff
@@ -0,0 +1,186 @@
+
+    this.targets_.set(target, false);
+
+    this.observer_.observe(target);
+
+    if (this.parsingTargets_ && !this.parsingTargets_.includes(target)) {
+      this.parsingTargets_.push(target);
+      this.checkParsing_();
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  unschedule(target) {
+    if (!this.targets_.has(target)) {
+      return;
+    }
+
+    console.log('DeferredBuilder: unschedule:', target.id);
+
+    this.targets_.delete(target);
+
+    this.observer_.unobserve(target);
+
+    if (this.parsingTargets_) {
+      removeItem(this.parsingTargets_, target);
+      this.checkParsing_();
+    }
+  }
+
+  /** @private */
+  docVisibilityChanged_() {
+    this.targets_.forEach((_, target) => this.maybeBuild_(target));
+  }
+
+  /** @private */
+  checkParsing_() {
+    const parsingTargets = this.parsingTargets_;
+    if (parsingTargets) {
+      for (let i = 0; i < parsingTargets.length; i++) {
+        const target = parsingTargets[i];
+        console.log(
+          'DeferredBuilder: parse complete?:',
+          target.id,
+          this.documentReady_,
+          hasNextNodeInDocumentOrder(target, this.ampdoc_.getRootNode())
+        );
+        if (
+          this.documentReady_ ||
+          hasNextNodeInDocumentOrder(target, this.ampdoc_.getRootNode())
+        ) {
+          parsingTargets.splice(i--, 1);
+
+          this.maybeBuild_(target);
+        }
+      }
+    }
+    if (this.documentReady_) {
+      this.parsingTargets_ = null;
+    }
+  }
+
+  /**
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   * @private
+   */
+  observed_(entries) {
+    for (let i = 0; i < entries.length; i++) {
+      const {target, isIntersecting} = entries[i];
+      console.log('DeferredBuilder: observed_:', target.id, isIntersecting);
+
+      if (!this.targets_.has(target)) {
+        continue;
+      }
+
+      this.targets_.set(target, isIntersecting);
+      if (isIntersecting) {
+        this.maybeBuild_(target);
+      }
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   * @private
+   */
+  maybeBuild_(target) {
+    const parsingTargets = this.parsingTargets_;
+    const parsed = !(parsingTargets && parsingTargets.includes(target));
+    const isIntersecting = this.targets_.get(target);
+    const toBuild =
+      parsed &&
+      isIntersecting &&
+      (this.ampdoc_.isVisible() ||
+        (this.ampdoc_.getVisibilityState() == VisibilityState.PRERENDER &&
+          target.prerenderAllowed()));
+    console.log('DeferredBuilder: build?:', target.id, toBuild, {
+      parsed,
+      isIntersecting,
+      dovVis: this.ampdoc_.getVisibilityState(),
+      prerender: target.prerenderAllowed(),
+    });
+    if (!toBuild) {
+      return;
+    }
+
+    this.unschedule(target);
+    // DO NOT SUBMIT: check the perform of this `setTimeout`. How else to ensure
+    // we do not over-build in a same task?
+    this.ampdoc_.win.setTimeout(() => target.build());
```

Comment:
`setTimeout` is **extremely** slow when called multiple times in a tick.

Reply:
Yeah. I'm still investigating this part. Not strictly necessary, but seemed like it'd be a missed opportunity to defer the work, but then still introduce a long task. Do you have any suggestions? I'd like to return control to the browser at least some of the time.